File: compiler/src/codegen/monomorphize.rs
use crate::{
    codegen::{CodegenCtx, CodegenResult, ir},
    database::{Db, NodeRef},
    typecheck::Substitutions,
    visit::{Defined, Definition},
};
use std::{
    collections::{BTreeMap, BTreeSet},
    mem,
};

#[derive(Debug, Default)]
pub struct MonomorphizeCtx {
    definitions: BTreeSet<ir::DefinitionKey>,
}

impl MonomorphizeCtx {
    pub fn get_or_insert(
        &mut self,
        definition: &NodeRef,
        substitutions: &Substitutions,
        bounds: BTreeMap<NodeRef, ir::Instance>,
        generic: bool,
        db: &Db,
    ) -> CodegenResult<ir::DefinitionKey> {
        let key = ir::DefinitionKey {
            node: definition.clone(),
            substitutions: convert_substitutions(substitutions, db)?,
            bounds,
        };

        if !generic {
            self.insert_monomorphized_key(key.clone());
        }

        Ok(key)
    }

    pub fn insert_monomorphized_key(&mut self, key: ir::DefinitionKey) {
        self.definitions.insert(key);
    }
}

fn convert_substitutions(
    substitutions: &Substitutions,
    db: &Db,
) -> CodegenResult<BTreeMap<NodeRef, ir::Type>> {
    substitutions
        .entries()
        .map(|(parameter, ty)| {
            Ok((
                parameter,
                ty.key(db)
                    .ok_or_else(|| anyhow::format_err!("cannot codegen type"))?,
            ))
        })
        .collect()
}

impl MonomorphizeCtx {
    pub fn monomorphize_definitions(
        &mut self,
        db: &mut Db,
    ) -> CodegenResult<impl Iterator<Item = (ir::DefinitionKey, Vec<ir::Instruction>)> + use<>>
    {
        let mut cache = BTreeMap::new();

        for key in mem::take(&mut self.definitions) {
            self.monomorphize_definition(&key, &mut cache, db)?;
        }

        Ok(cache
            .into_iter()
            .map(|(key, instructions)| (key, instructions.unwrap())))
    }

    fn monomorphize_definition(
        &mut self,
        key: &ir::DefinitionKey,
        cache: &mut BTreeMap<ir::DefinitionKey, Option<Vec<ir::Instruction>>>,
        db: &mut Db,
    ) -> CodegenResult {
        if cache.contains_key(key) {
            return Ok(());
        }

        cache.insert(key.clone(), None);

        let Defined(definition) = db
            .get(&key.node)
            .ok_or_else(|| anyhow::format_err!("no definition for {key:?}"))?;

        let body = match definition {
            Definition::Constant(definition) => definition.value,
            Definition::Instance(definition) => definition.value,
            _ => None,
        }
        .ok_or_else(|| anyhow::format_err!("definition has no value"))?;

        let mut ctx = CodegenCtx::new(db);
        ctx.codegen(&body)?;
        ctx.instruction(ir::Instruction::Return {
            value: body.clone(),
        });

        let mut instructions = ctx.pop_instructions();
        for instruction in &mut instructions {
            instruction.traverse_mut(&mut |instruction| {
                if let ir::Instruction::Value { value, .. } = instruction {
                    match value {
                        ir::Value::Constant(constant_key) => {
                            self.monomorphize_key(
                                constant_key,
                                &key.substitutions,
                                &key.bounds,
                                cache,
                                db,
                            )?;
                        }
                        ir::Value::Bound(bound) => {
                            let Some(ir::Instance::Definition(mut resolved_key)) =
                                key.bounds.get(bound).cloned()
                            else {
                                return Err(anyhow::format_err!("bound {bound:?} not resolved"));
                            };

                            self.monomorphize_key(
                                &mut resolved_key,
                                &key.substitutions,
                                &key.bounds,
                                cache,
                                db,
                            )?;

                            *value = ir::Value::Constant(resolved_key);
                        }
                        _ => {}
                    }
                }

                Ok(())
            })?;
        }

        cache.get_mut(key).unwrap().replace(instructions);

        Ok(())
    }

    fn monomorphize_key(
        &mut self,
        key: &mut ir::DefinitionKey,
        substitutions: &BTreeMap<NodeRef, ir::Type>,
        bounds: &BTreeMap<NodeRef, ir::Instance>,
        cache: &mut BTreeMap<ir::DefinitionKey, Option<Vec<ir::Instruction>>>,
        db: &mut Db,
    ) -> CodegenResult {
        for ty in key.substitutions.values_mut() {
            self.monomorphize_type(ty, substitutions)?;
        }

        for instance in key.bounds.values_mut() {
            self.monomorphize_instance(instance, substitutions, bounds, cache, db)?;
        }

        self.monomorphize_definition(key, cache, db)?;

        Ok(())
    }

    fn monomorphize_type(
        &mut self,
        ty: &mut ir::Type,
        substitutions: &BTreeMap<NodeRef, ir::Type>,
    ) -> CodegenResult {
        let mut success = true;
        ty.traverse_mut(&mut |ty| {
            if !success {
                return;
            }

            if let ir::Type::Parameter(ref parameter) = *ty {
                if let Some(substitution) = substitutions.get(parameter) {
                    *ty = substitution.clone();
                } else {
                    success = false;
                }
            }
        });

        if !success {
            return Err(anyhow::format_err!("could not monomorphize {ty:?}"))?;
        }

        Ok(())
    }

    fn monomorphize_instance(
        &mut self,
        instance: &mut ir::Instance,
        substitutions: &BTreeMap<NodeRef, ir::Type>,
        bounds: &BTreeMap<NodeRef, ir::Instance>,
        cache: &mut BTreeMap<ir::DefinitionKey, Option<Vec<ir::Instruction>>>,
        db: &mut Db,
    ) -> CodegenResult {
        match instance {
            ir::Instance::Bound(bound) => {
                *instance = bounds
                    .get(bound)
                    .ok_or_else(|| anyhow::format_err!("no bound for {bound:?}"))?
                    .clone();
            }
            ir::Instance::Definition(instance_key) => {
                self.monomorphize_key(instance_key, substitutions, bounds, cache, db)?;
            }
        }

        Ok(())
    }
}

File: compiler/src/codegen/Monomorphize.swift
import OrderedCollections

class MonomorphizeContext {
    let db: DB
    private var definitions: OrderedSet<IR.DefinitionKey> = []

    init(db: DB) { self.db = db }

    func getOrInsert(
        definition: Node,
        substitutions: Substitutions,
        bounds: OrderedDictionary<Node, IR.Instance>,
        isGeneric: Bool,
    ) throws -> IR.DefinitionKey {
        let key = IR.DefinitionKey(
            node: definition,
            substitutions: try convertSubstitutions(substitutions),
            bounds: bounds,
        )

        if !isGeneric { self.insert(monomorphized: key) }

        return key
    }

    func insert(monomorphized key: IR.DefinitionKey) { self.definitions.append(key) }

    private func convertSubstitutions(_ substitutions: Substitutions) throws -> OrderedDictionary<
        Node, IR.`Type`
    > {
        var converted: OrderedDictionary<Node, IR.`Type`> = [:]
        for (parameter, type) in substitutions {
            converted[parameter] = try self.db.irType(for: type)
        }

        return converted
    }

    public func monomorphizeDefinitions() throws -> some Sequence<(IR.DefinitionKey, IR.Definition)>
    {
        let cache = Cache()
        for key in self.definitions { try self.monomorphizeDefinition(key: key, cache: cache) }

        return cache.inner.lazy.map { key, slot in (key, slot.definition!) }
    }

    private class Cache {
        enum Slot {
            case empty
            case definition(IR.Definition)

            var definition: IR.Definition? {
                switch self {
                case .empty: return nil
                case .definition(let definition): return definition
                }
            }
        }

        var inner: OrderedDictionary<IR.DefinitionKey, Slot> = [:]
    }

    private func monomorphizeDefinition(key: IR.DefinitionKey, cache: Cache) throws {
        guard cache.inner[key] == nil else { return }

        cache.inner[key] = .empty

        guard let definition = self.db[key.node, Defined.self]?.definition else {
            throw CodegenError("no definition for \(key)")
        }

        let body: Node? =
            switch definition {
            case let definition as ConstantDefinition: definition.value
            case let definition as InstanceDefinition: definition.value
            default: nil
            }

        guard let body else { throw CodegenError("definition \(key) has no value") }

        let context = CodegenContext(db: self.db)
        try context.codegen(body)
        context.instruction(.return(value: body))

        var instructions = context.popInstructions()
        var types: OrderedDictionary<Node, IR.`Type`> = [:]
        for index in instructions.indices {
            instructions[index] = try instructions[index]
                .traverse { instruction in
                    try instruction.forEachNode { node in
                        types[node] = try self.db.irType(for: .node(node))
                            .substitute(from: key.substitutions)
                    }

                    var instruction = instruction
                    if case .value(let node, let value) = instruction {
                        switch value {
                        case .constant(let constantKey):
                            instruction = .value(
                                node: node,
                                value: .constant(
                                    try self.monomorphizeKey(
                                        constantKey,
                                        substitutions: key.substitutions,
                                        bounds: key.bounds,
                                        cache: cache,
                                    )
                                ),
                            )
                        case .bound(let bound):
                            guard case .definition(let resolvedKey) = key.bounds[bound] else {
                                throw CodegenError("bound \(bound) not resolved")
                            }

                            instruction = .value(
                                node: node,
                                value: .constant(
                                    try self.monomorphizeKey(
                                        resolvedKey,
                                        substitutions: key.substitutions,
                                        bounds: key.bounds,
                                        cache: cache,
                                    )
                                ),
                            )
                        default: break
                        }
                    }

                    return instruction
                }
        }

        cache.inner[key] = .definition(
            .init(type: types[body], instructions: instructions, types: types)
        )
    }

    private func monomorphizeKey(
        _ key: IR.DefinitionKey,
        substitutions: OrderedDictionary<Node, IR.`Type`>,
        bounds: OrderedDictionary<Node, IR.Instance>,
        cache: Cache,
    ) throws -> IR.DefinitionKey {
        var key = key

        for (parameter, type) in key.substitutions {
            key.substitutions[parameter] = try self.monomorphizeType(
                type,
                substitutions: substitutions,
            )
        }

        for (boundNode, instance) in key.bounds {
            key.bounds[boundNode] = try self.monomorphizeInstance(
                instance,
                substitutions: substitutions,
                bounds: bounds,
                cache: cache,
            )
        }

        try self.monomorphizeDefinition(key: key, cache: cache)

        return key
    }

    private func monomorphizeType(
        _ type: IR.`Type`,
        substitutions: OrderedDictionary<Node, IR.`Type`>,
    ) throws -> IR.`Type` {
        let root = type
        return try type.traverse { type in
            if case .parameter(let parameter) = type {
                guard let substitution = substitutions[parameter] else {
                    throw CodegenError(
                        "could not monomorphize \(root) with substitutions \(substitutions)"
                    )
                }

                return substitution
            }

            return type
        }
    }

    private func monomorphizeInstance(
        _ instance: IR.Instance,
        substitutions: OrderedDictionary<Node, IR.`Type`>,
        bounds: OrderedDictionary<Node, IR.Instance>,
        cache: Cache,
    ) throws -> IR.Instance {
        switch instance {
        case .bound(let bound):
            guard let instance = bounds[bound] else { throw CodegenError("no bound for \(bound)") }
            return instance
        case .definition(let key):
            return .definition(
                try self.monomorphizeKey(
                    key,
                    substitutions: substitutions,
                    bounds: bounds,
                    cache: cache,
                )
            )
        }
    }
}
